### PR TITLE
Improve performance with optimized common method/iter implementation shortcuts

### PR DESCRIPTION
**Pull request by hmdne**

### CODE REVIEW

File: opal/corelib/runtime.js [hunk 1]
```diff
@@ -2809,7 +2809,30 @@
     return $truthy(arg['$!']());
   }
 
+  // Optimized function generator for simple functions
+  function $value_fun(arg) {
```
Comment (hmdne) [inline opal/corelib/runtime.js:2813]: Think about the names...

**Comment by hmdne** [comment]
I added optimizations for code like:

```
def some_delegator
  @ivar.method
end
```

I am afraid this may break backtraces. Do we gain something from that? Perhaps we can introduce something like optimization levels, eg.:

```
-O0: return output maximally compatible with MRI
-O1: (default) return output maximally compatible with ISO Ruby (MRuby)
-O2: sacrifice potential stacktraces
```

I know `-O` is used, so it's just a food for thought for now.

**Comment by hmdne** [comment]
For a code like:

```ruby
def x
  y
end

def y
  z
end

def z
  error
end

x
```

We get this stacktrace:

image

I wonder if this is acceptable, though the gains are quite clear.

File: lib/opal/nodes/iter.rb [hunk 2]
```diff
@@ -1,52 +1,19 @@
 # frozen_string_literal: true
 
-require 'opal/nodes/node_with_args'
+require 'opal/nodes/def'
```
Comment (elia) [inline lib/opal/nodes/iter.rb:3]: Not sure why this change, looks like we're still inheriting from NodeWithArgs, is there another reason?

**Review by elia — APPROVED** [review]
Just one nit but 👍👍👍 overall!

File: lib/opal/nodes/iter.rb (hunk 2 above)
Comment (hmdne) [reply lib/opal/nodes/iter.rb:3]: My mistake. I wanted to do one thing inheriting another.

**Review by elia — COMMENTED** [review]
After discussing this together we agreed to put this behind an optimisation flag, e.g. `--optimize-defs` due to the loss of precision in the stacktrace

**Comment by hmdne** [comment]
We agreed to get a limited version for v1.5, the part that doesn't mess up backtraces. For v1.6 we will ship a massively improved changes.
